pkg/common/cns-lib/volume/manager.go

```diff
@@ -2112,6 +2112,56 @@ func (m *defaultManager) deleteSnapshotWithImprovedIdempotencyCheck(
 	// Get the taskInfo
 	deleteSnapshotsTaskInfo, err := cns.GetTaskInfo(ctx, deleteSnapshotTask)
 	if err != nil || deleteSnapshotsTaskInfo == nil {
+		if err != nil && cnsvsphere.IsManagedObjectNotFound(err, deleteSnapshotTask.Reference()) {
+			log.Infof("Snapshot %q on volume %q might have already been deleted "+
+				"with the error %v. Calling CNS QuerySnapshots API to confirm it", snapshotID, volumeID,
+				spew.Sdump(err))
+
+			// To double check whether the DeleteSnapshot task is completed in the underlying vSphere.
+			snapshotQueryFilter := cnstypes.CnsSnapshotQueryFilter{
+				SnapshotQuerySpecs: []cnstypes.CnsSnapshotQuerySpec{
+					{VolumeId: cnstypes.CnsVolumeId{Id: volumeID}, SnapshotId: &cnstypes.CnsSnapshotId{Id: snapshotID}},
+				},
+				Cursor: &cnstypes.CnsCursor{Offset: 0, Limit: 1},
+			}
+
+			if _, querySnapshotErr := m.QuerySnapshots(ctx, snapshotQueryFilter); querySnapshotErr != nil {
+				// Use a separate error variable, querySnapshotErr, here to prevent the error from being propagated
+				if cnsvsphere.IsCnsSnapshotNotFoundError(querySnapshotErr) {
+					// Confirm the the snapshot is deleted successfully in vSphere
+					log.Infof("Snapshot %q on volume %q might have already been deleted "+
+						"with the error %v. Ignore the error for DeleteSnapshot", snapshotID, volumeID,
+						spew.Sdump(querySnapshotErr))
+
+					var taskInvocationTimestamp metav1.Time
+					var activationId string
+					if volumeOperationDetails != nil {
+						taskInvocationTimestamp = volumeOperationDetails.OperationDetails.TaskInvocationTimestamp
+						activationId = volumeOperationDetails.OperationDetails.OpID
+					}
+
+					if m.idempotencyHandlingEnabled {
+						// Create the volumeOperationDetails object for persistence
+						volumeOperationDetails = createRequestDetails(instanceName, "", "", 0,
+							taskInvocationTimestamp, deleteSnapshotTask.Reference().Value,
+							activationId, taskInvocationStatusSuccess, "")
+					}
+
+					log.Infof("DeleteSnapshot: Snapshot %q on volume %q is deleted successfully. opId: %q", snapshotID,
+						volumeID, activationId)
+
+					return nil
+				}
+
+				log.Infof("Cannot ignore the ManagedObjectNotFound error since the CNS QuerySnapshots API call "+
+					"failed with error %v", spew.Sdump(querySnapshotErr))
+			} else {
+				log.Infof("Cannot ignore the ManagedObjectNotFound error since the CNS QuerySnapshots API call "+
+					"confirms the snapshot %q on volume %q still exists. "+
+					"It implies a bug in the underlying vSphere CNS layer", snapshotID, volumeID)
+			}
```

Comment:
Maybe fewer words, I don't think it's good to mention the word "bug" in logs

Reply:
Should this be a warning?

Reply:
Same here. Maybe shorter message.

Reply:
Reword the log message even the code structure. Please take a look.